### ISSUE TITLE
Determine the scope and limitations of OpenKiln

### ISSUE THREAD

**Issue by austinjkee** [austinjkee/OpenKiln#1]
What specific use cases are we planning on targeting?  More importantly, what features do we plan on ignoring in the original design (e.g.: undo, texturing, lighting, etc.)?

**Comment by charliedraper** [comment]
As far as use-cases, some of the ideas that I had from the end-user's perspective are:
- Selects Toolbar Icon
- Sets Mode (for drawing or moving/rotating)
- Sets Pen Thickness
- Sets Color
- Opens/Closes Project
- Saves/Deletes Project

In terms of features we could pull out of this, I think some important ones are:
 - the toolbar, which contains pen customization and the ability to select 3D (and 2D?) shapes, an undo button, and a button for altering between drawing and moving around the viewing angle
- the canvas, which should exist in a 3D coordinate plane. The coordinates could also be changed between Cartesian and polar and could be toggled on and off
- an options menu for opening, closing, saving, and deleting the project, as well as for other customizations to the application

As far as the more technical components of a drawing application like texturing, lighting, and shading, I don't know a lot about how these are generally implemented but I'd be happy to try to incorporate them if anybody else has a good grasp on how we could work them in

**Comment by odunbar3** [comment]
Maybe we should do an activity diagram for creating a simple drawing or adding a shape to the canvas?

I don't know if this is what you had in mind but some 3D  drawing apps have a measurement functionality to model things in real life or send them off to a laser cutter or 3D printer. We could implement the ability to use measurements but not necessarily use the measurements ourselves.

**Comment by charliedraper** [comment]
Yeah I like that idea

**Comment by s-wheeler** [comment]
Are the use cases supposed to be for the end user or a programmer using the framework?

**Comment by s-wheeler** [comment]
And same for activity and sequence diagrams

**Comment by austinjkee** [comment]
The use cases are supposed to be for a programmer, as we're just creating a framework, not an application.  But considering how someone would use it extends to their end users as well; we have to think about what applications would be built with our framework.

**Comment by charliedraper** [comment]
Yeah when I asked Ferby in class last week she said to think from the perspective of the end-user because that will give you more ideas about how to implement the framework

**Comment by austinjkee** [comment]
We should probably have a way to import and export to different formats from the internal format that we choose.

For measurements, I like the idea of adding a way to map pixels to real units.  Most existing 3D drawing applications internally use dimensional ratios between axes to map to real world measurements.  Internally, we'd use orthogonal coordinates and then have to convert to more user-friendly coordinate systems like Cartesian and polar, and definitely barycentric coordinates for export formats. 

We could also leave a prototype interface exposed for the coordinate system so that developers may use more niche systems if their application requires it (e.g.: Grassmannian for n-dimension representations, etc.) or simpler performance optimized systems that don't require all of the generalized features.

**Comment by odunbar3** [comment]
For the activity and sequence diagrams, they are more so from the end user perspective i think. I was thinking about using them to help us figure out which classes and objects we needed.

I think it is a really good idea to all the developer to decide how much of functionality is required for their system while having the basics for measurements in our map. 

Another use case diagram we probably need is the ability to choose and draw a 3D shape like a sphere or a pyramid and one for the ability to draw a free form shape. In Paint 3D, there is the ability to draw 3D and 2D shapes. Do we only want to be able to draw 3D shapes or to also draw 2D shapes?
